### PR TITLE
Support session encryption

### PR DESCRIPTION
**Pull request by kkadosh**

### CODE REVIEW

File: src/php_snuffleupagus.h [hunk 1]
```diff
@@ -65,6 +67,10 @@ bool is_config_valid;
 HashTable *disabled_functions_hook;
 HashTable *sp_internal_functions_hook;
 HashTable *sp_eval_blacklist_functions_hook;
+void *s_module;
+void *s_original_mod;
+int (*old_s_read)(PS_READ_ARGS);
```
Comment (jvoisin) [inline src/php_snuffleupagus.h]: This could be a static variable in `sp_session.c`

File: src/sp_cookie_encryption.c [hunk 2]
```diff
@@ -36,11 +35,6 @@ int decrypt_cookie(zval *pDest, int num_args, va_list args,
   return decrypt_zval(pDest, cookie->simulation, hash_key);
 }
 
-/*
```
Comment (jvoisin) [inline src/sp_cookie_encryption.c]: Why did you remove this comment?

File: src/sp_crypt.c [hunk 3]
```diff
@@ -100,37 +97,39 @@ int decrypt_zval(zval *pDest, bool simulation, zend_hash_key *hash_key) {
...
+** operation).
+*/
 zend_string *encrypt_zval(char *data, unsigned long long data_len) {
   const size_t encrypted_msg_len = crypto_secretbox_ZEROBYTES + data_len + 1;
+  // FIXME : We know that this len is too long
   const size_t emsg_and_nonce_len =
       encrypted_msg_len + crypto_secretbox_NONCEBYTES;
 
   unsigned char key[crypto_secretbox_KEYBYTES] = {0};
-  unsigned char nonce[crypto_secretbox_NONCEBYTES] = {0};
```
Comment (jvoisin) [inline src/sp_crypt.c]: `{0}` would be alright ;)

File: src/sp_session.c [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+#endif
+#else
+static php_ps_globals *session_globals = NULL;
+#define SESSION_G(v) (ps_globals.v)
+#endif
+
+static int (*previous_sessionRINIT)(INIT_FUNC_ARGS) = NULL;
+static ZEND_INI_MH((*old_OnUpdateSaveHandler)) = NULL;
+
+// PS_READ_ARGS => void **mod_data, zend_string *key, zend_string **val,
```
Comment (jvoisin) [inline src/sp_session.c]: Either trash this comment, or replace `PS_READ_ARGS` by the arguments in the function prototype.

File: src/sp_session.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  int r = SNUFFLEUPAGUS_G(old_s_read)(mod_data, key, val, maxlifetime);
+  if (r == SUCCESS && SNUFFLEUPAGUS_G(config).config_session->encrypt &&
+      val != NULL && *val != NULL && ZSTR_LEN(*val)) {
+    zend_string *orig_val = *val;
+    zval val_zval;
+    ZVAL_PSTRINGL(&val_zval, ZSTR_VAL(*val), ZSTR_LEN(*val));
+
+    int ret = decrypt_zval(
+        &val_zval, SNUFFLEUPAGUS_G(config).config_session->simulation,
+        NULL);  // NULL for the moment
```
Comment (jvoisin) [inline src/sp_session.c]: What does `// NULL for the moment` mean?

File: src/sp_session.c [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  mod->s_read = sp_hook_s_read;
+
+  SNUFFLEUPAGUS_G(old_s_write) = mod->s_write;
+  mod->s_write = sp_hook_s_write;
+
+  SESSION_G(mod) = mod;
+}
+
+static PHP_INI_MH(sp_OnUpdateSaveHandler) {
+  if (stage == PHP_INI_STAGE_RUNTIME &&
```
Comment (jvoisin) [inline src/sp_session.c]: What is the point of this? I don't get where the "user" string is coming from.

File: src/sp_cookie_encryption.c (hunk 2 above)
Comment (kkadosh) [reply src/sp_cookie_encryption.c]: I have move it into `sp_crypt.c` .

File: src/sp_config_keywords.c [hunk 7]
```diff
@@ -60,6 +60,46 @@ static int parse_enable(char *line, bool *restrict retval,
   return ret;
 }
 
+int parse_session(char *line) {
+  sp_config_session *session =
+      pecalloc(sizeof(sp_config_session), 1, 1);
```
Comment (jvoisin) [inline src/sp_config_keywords.c]: Can't the last argument be zero instead of one?

File: src/sp_config_keywords.c [hunk 8]
```diff
@@ -60,6 +60,46 @@ static int parse_enable(char *line, bool *restrict retval,
...
+      {parse_empty, SP_TOKEN_SIMULATION, &(session->simulation)},
+      {0}};
+  int ret = parse_keywords(sp_config_funcs_session_encryption, line);
+  if (0 != ret) {
+    return ret;
+  }
+  if (session->encrypt) {
+    if (0 == (SNUFFLEUPAGUS_G(config).config_snuffleupagus->cookies_env_var)) {
+      sp_log_err(
+          "config",
```
Comment (jvoisin) [inline src/sp_config_keywords.c]: It's not `config`, but `session_encryption` I guess

File: src/sp_config_keywords.c [hunk 9]
```diff
@@ -60,6 +60,46 @@ static int parse_enable(char *line, bool *restrict retval,
...
+          "You're trying to use the session cookie encryption feature"
+          "on line %zu without having set the `.cookie_env_var` option in"
+          "`sp.global`: please set it first.",
+          sp_line_no);
+      return -1;
+    } else if (0 ==
+               (SNUFFLEUPAGUS_G(config).config_snuffleupagus->encryption_key)) {
+      sp_log_err(
+          "config",
```
Comment (jvoisin) [inline src/sp_config_keywords.c]: Same remark as previous comment :)
